File: ProjetoFuncional/Criaturas.hs
module Criaturas where

import Inventario
import Itens
import Mundo
import Personagem
import Text.Printf
import Util
import Mundo (setArmadilha)

-- Besta Implementacao
ataqueBesta :: Status -> [Item] -> Status
ataqueBesta status itens
    | elem counterBesta itens = (setDiario (mensagemEscape) personagem, mundo) 
    | otherwise = (setVida (life - 25) personagemEscreverDiario, mundo)
    where personagem = fst status
          life = vida personagem
          mundo = snd status
          mensagemAtaque = diario personagem ++ "Voce foi atacado pela Besta\n"
          mensagemEscape = mensagemAtaque ++ "Mas conseguiu evita-la\n"
          personagemEscreverDiario = setDiario mensagemAtaque personagem

encontroBesta :: Status -> Int -> Bool
encontroBesta status idLocal 
    | mod day 3 == 0 && momento == Manha && idLocal == 2 = True
    | mod day 2 == 0 && momento == Manha && idLocal == 3 = True
    | mod day 1 == 0 && momento == Manha && idLocal == 4 = True
    | mod day 2 == 0 && momento == Tarde && idLocal == 3 = True
    | mod day 1 == 0 && momento == Tarde && idLocal == 4 = True
    | momento == Noite && idLocal == 2 = True
    | otherwise = False
    where mundo = snd status
          day = dia mundo
          momento = momentoDia mundo

-- Perseguidor Implementacao
cairArmadilha :: Status -> [Item] -> Status
cairArmadilha status itens
    | elem counterArmadilha itens = (setDiario (mensagemEscape) personagem, mundoLivre) 
    | otherwise = (setVida (life - 25) personagemEscreverDiario, mundoLivre)
    where personagem = fst status
          life = vida personagem
          mundo = snd status
          mundoLivre = setArmadilha [Nenhuma, Nenhuma, Nenhuma, Nenhuma, Nenhuma] mundo
          mensagemAtaque = diario personagem ++ "Voce caiu numa armadilha\n"
          mensagemEscape = mensagemAtaque ++ "Mas conseguiu evita-la\n"
          personagemEscreverDiario = setDiario mensagemAtaque personagem

encontroArmadilha :: Status -> Int -> Bool
encontroArmadilha status idLocal = (armadilhas !! (idLocal - 1)) == Armada
    where mundo = snd status
          armadilhas = locaisArmadilha mundo

espantarPerseguidor :: Status -> Status
espantarPerseguidor status = (personagemEscreverDiario, mundoLivre)
    where personagem = fst status
          mensagem = diario personagem ++ "Seu Perseguidor estava colocando armadilhas\nVoce o espantou\n"
          personagemEscreverDiario = setDiario mensagem personagem
          mundo = snd status
          mundoLivre = setArmadilha [Nenhuma, Nenhuma, Nenhuma, Nenhuma, Nenhuma] mundo

encontroPerseguidor :: Status -> Int -> Bool
encontroPerseguidor status idLocal = armadilhas !! (idLocal - 1) == Armando
    where mundo = snd status
          armadilhas = locaisArmadilha mundo

atualizarArmadilha :: Status -> Int -> Status
atualizarArmadilha status idLocal 
    | elem Armada armadilhas = status
    | elem Armando armadilhas = (personagem, setArmadilha (armarArmadilha armadilhas) mundo)
    | otherwise = (personagem, setArmadilha armando mundo)
    where personagem = fst status
          mundo = snd status 
          armadilhas = locaisArmadilha mundo
          armando = (take (idLocal-1) armadilhas) ++ [Armando] ++ (drop idLocal armadilhas)

armarArmadilha :: [StatusArmadilha] -> [StatusArmadilha]
armarArmadilha [] = []
armarArmadilha (head:tail) = if head == Armando 
                                then (Armada:tail)
                                else (head:armarArmadilha tail)

-- Robo Implementacao
ataqueRoboBaseMilitar :: Status -> [Item] -> Status
ataqueRoboBaseMilitar status itens
    | elem counterRobo itens = (setDiario (mensagemEscape) personagem, mundo) 
    | otherwise = (setVida (life - 25) personagemEscreverDiario, mundo)
    where personagem = fst status
          life = vida personagem
          mundo = snd status
          mensagemAtaque = diario personagem ++ "Voce foi atacado pelo Robo\n"
          mensagemEscape = mensagemAtaque ++ "Mas conseguiu evita-lo\n"
          personagemEscreverDiario = setDiario mensagemAtaque personagem

encontroRoboBaseMilitar :: Status -> Int -> Bool
encontroRoboBaseMilitar status idLocal = (not (mod day 2 == 0)) && (idLocal == 1)
    where mundo = snd status
          day = dia mundo

File: ProjetoLogico/criaturas.pl
% BESTA IMPL

encontroBesta(Dia, MomentoDia, IdLocal) :- (Dia mod 3) =:= 0, MomentoDia = "MANHA", IdLocal =:= 2.
encontroBesta(Dia, MomentoDia, IdLocal) :- (Dia mod 2) =:= 0, MomentoDia = "MANHA", IdLocal =:= 3.
encontroBesta(Dia, MomentoDia, IdLocal) :- (Dia mod 1) =:= 0, MomentoDia = "MANHA", IdLocal =:= 4.
encontroBesta(Dia, MomentoDia, IdLocal) :- (Dia mod 3) =:= 0, MomentoDia = "TARDE", IdLocal =:= 4.
encontroBesta(Dia, MomentoDia, IdLocal) :- (Dia mod 3) =:= 0, MomentoDia = "TARDE", IdLocal =:= 3.
encontroBesta(_, MomentoDia, IdLocal) :- MomentoDia = "NOITE", IdLocal =:= 2.

ataqueBesta(Vida, Vida2) :- writeln("Voce foi atacado pela Besta\n"), Vida2 is Vida - 25.

% PERSEGUIDOR IMPL

encontroArmadilha(Armadilhas, IdLocal) :- I is IdLocal - 1, nth0(I, Armadilhas, "ARMADA").

cairArmadilha(Vida, Vida2, ["NENHUMA", "NENHUMA", "NENHUMA", "NENHUMA", "NENHUMA"]) :- writeln("Voce caiu numa armadilha\n"), Vida2 is Vida - 25.

encontroPerseguidor(Armadilhas, IdLocal) :- I is IdLocal - 1, nth0(I, Armadilhas, "ARMANDO").

espantarPerseguidor(["NENHUMA", "NENHUMA", "NENHUMA", "NENHUMA", "NENHUMA"]) :- writeln("Seu Perseguidor estava colocando armadilhas\nVoce o espantou\n").

atualizarArmadilha(Armadilhas, _, Armadilhas2) :- member("ARMADA", Armadilhas), Armadilhas2 = Armadilhas, !.
atualizarArmadilha(Armadilhas, _, Armadilhas2) :- member("ARMANDO", Armadilhas), armarArmadilha(Armadilhas, Armadilhas2), !.
atualizarArmadilha(Armadilhas, IdLocal, Armadilhas2) :- armandoArmadilha(IdLocal, Armadilhas, Armadilhas2).

armandoArmadilha(1, [_|C], ["ARMANDO"|C]):- !.
armandoArmadilha(I, [Y|C], [Y|D]) :- I2 is I - 1, armandoArmadilha(I2, C, D).

armarArmadilha(["ARMANDO"|C], ["ARMADA"|C]):- !.
armarArmadilha([Y|C], [Y|D]) :- armarArmadilha(C, D).

% ROBO IMPL

encontroRoboBaseMilitar(Dia, MomentoDia, IdLocal) :- (Dia mod 2) =:= 0, MomentoDia = "MANHA", IdLocal =:= 1.
encontroRoboBaseMilitar(Dia, MomentoDia, IdLocal) :- (Dia mod 2) =:= 0, MomentoDia = "TARDE", IdLocal =:= 1.

ataqueRoboBaseMilitar(Vida, Vida2) :- writeln("Voce foi atacado pelo Robo\n"), Vida2 is Vida - 25.

roboBarulhos(Dia) :- member(Dia, [3, 5, 7, 9, 11, 13]) -> writeln("Escuta-se ao longe barulhos de engrenagem"); write("").
        

/*roboDestruirLocal :: Int -> Status -> Status
roboDestruirLocal dia status
  | dia == 5 = (personagem, roboAfetarArea 4 mundo)
  | dia == 9 = (personagem, roboAfetarArea 3 mundo)
  | dia == 13 = (personagem, roboAfetarArea 2 mundo)
  | otherwise = status
  where mundo = snd status
        personagem = fst status */

